Support for Traefik v2

simonkrenger opened the pull request:
These commits add support for the new Traefik v2 configuration format. Starting
with Traefik 2.0, there are now two parts of a Traefik configuration, static
and dynamic configuration. This means the configuration format changed
considerably from a relatively simple format to something more complex.

In `config.js`, updated the latest version of Traefik. In `traefik.hbs`, added
the complete template for the new configuration format, as found in the Traefik
documentation. Especially relevant are the following documentation sections:

- [ TLS Configuration Is Now Dynamic, per Router ]
  https://docs.traefik.io/migration/v1-to-v2/#tls-configuration-is-now-dynamic-per-router

- [ HTTP to HTTPS Redirection Is Now Configured on Routers ]
  https://docs.traefik.io/migration/v1-to-v2/#http-to-https-redirection-is-now-configured-on-routers

Fixes #66

april commented:
Is it necessary to have the regular and "my" section, for Traefix 2.0?

simonkrenger commented:
In general, yes, there are now multiple configuration sections necessary to configure the necessary options. As Traefik acts as a reverse proxy, one needs to configure "EntryPoints", "Routers", "TLS Options" and "Middlewares".

However, the configuration could certainly be trimmed down - that is a question on what the Configuration Generator should provide. Should it provide a full, ready-to-go configuration (as the Traefik documentation provides and how it is currently implemented) or just the configuration sections relevant for TLS configuration to make it more generic?

An alternative could be that we generate the "my" sections commented-out, e.g. like so to suggest that the user needs to provide his own values here:

```
  [http.routers.router-secure]
    # rule = "Host(`foo.com`)"
    # service = "my-service"
    entrypoints = ["web-secure"]
    [http.routers.router-secure.tls]
      options = "TLSOptions"
```

april commented:
In general, I try to stick pretty close to the configuration file the provider offers, and it's rarely enough to have a fully configured setup.

Mostly I am just covering the minimum amount to get TLS configured correctly - anything else adds to a higher level of maintenance. As such, I currently have (locally) something like this:

```
# generated {{output.date}}, {{{output.link}}}
{{#if (minver "2.0.0" form.serverVersion)}}
{{! traefik 2.0 has a very different configuration style}}
[http.routers]
  [http.routers.Router-1]
    rule = "Host(`bar.com`)"
    service = "service-id"

    [http.routers.Router-1.tls]
      options = "{{form.config}}"
{{#if form.hsts}}

[http.middlewares]
  [http.middlewares.redirect.redirectScheme]
    scheme = "https"
{{/if}}

# due to Go limitations, it is highly recommended that you use an ECDSA
# certificate, or you may experience compatibility issues
[[tls.certificates]]
  certFile = "/path/to/signed_cert_plus_intermediates"
  keyFile = "/path/to/private_key"

[tls.options]
  [tls.options.{{form.config}}]
    minVersion = "{{{replace output.protocols.[0] "TLSv1." "VersionTLS1"}}}"
    {{#if output.ciphers.length}}
    cipherSuites = [
    {{#each output.ciphers}}
      "{{this}}"{{#unless @last}},{{/unless}}
    {{/each}}
    ]
      {{/if}}
{{else}}
defaultEntryPoints = ["http", "https"]

[entryPoints]
{{#if form.hsts}}
  [entryPoints.http]
  address = ":80"
    [entryPoints.http.redirect]
    entryPoint = "https"

{{/if}}
  [entryPoints.https]
  address = ":443"
    [entryPoints.https.tls]
      {{#if (eq output.protocols.[0] "TLSv1")}}
      minVersion = "VersionTLS10"
      {{else}}
      minVersion = "{{{replace output.protocols.[0] "TLSv1." "VersionTLS1"}}}"
      {{/if}}
      {{#if output.ciphers.length}}
      cipherSuites = [
      {{#each output.ciphers}}
        "{{this}}"{{#unless @last}},{{/unless}}
      {{/each}}
      ]
      {{/if}}

      # due to Go limitations, it is highly recommended that you use an ECDSA
      # certificate, or you may experience compatibility issues
      [[entryPoints.https.tls.certificates]]
      certFile = "/path/to/signed_cert_plus_intermediates"
      keyFile = "/path/to/private_key"
{{/if}}
```

Which produces something like this:

```
# generated 2020-01-16, http://localhost:5500/#server=traefik&server-version=2.1&config=intermediate
[http.routers]
  [http.routers.Router-1]
    rule = "Host(`bar.com`)"
    service = "service-id"

    [http.routers.Router-1.tls]
      options = "intermediate"

[http.middlewares]
  [http.middlewares.redirect.redirectScheme]
    scheme = "https"

# due to Go limitations, it is highly recommended that you use an ECDSA
# certificate, or you may experience compatibility issues
[[tls.certificates]]
  certFile = "/path/to/signed_cert_plus_intermediates"
  keyFile = "/path/to/private_key"

[tls.options]
  [tls.options.intermediate]
    minVersion = "VersionTLS12"
    cipherSuites = [
      "TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384",
      "TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384",
      "TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256",
      "TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256",
      "TLS_ECDHE_ECDSA_WITH_CHACHA20_POLY1305",
      "TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305"
    ]
```

Does this run okay without the `default` setting? I don't think it's necessary, and I like to not specify things whenever possible, especially if they could affect other parts o the system.

(btw, thanks for the HSTS thing, I didn't know that)

april commented:
It also looks like this adds the HSTS header, according to their docs:

```
[http.middlewares]
  [http.middlewares.testHeader.headers]
    [http.middlewares.testHeader.headers.customResponseHeaders]
        Strict-Transport-Security = "max-age={{output.hstsMaxAge}}"
```

But it's not entirely clear what `testHeader` is supposed to be. Is that the router name, e.g. `Router-1`? Thanks!

(btw, are you on Twitter? I'd love to credit you once I get this in there)

simonkrenger commented:
Thanks for the valuable feedback.

Covering the minimum amount to get TLS configured correctly definitely makes sense. I have updated the template by removing large parts of it and also incorporating most of your changes. The commits are yet untested, so I'll need to do the testing there before we can merge. Please allow me some time to do that.

> It also looks like this adds the HSTS header, according to their docs:

Thanks for catching that. I added this to the template.

> But it's not entirely clear what testHeader is supposed to be. Is that the router name, e.g. Router-1? Thanks!

That is the name of the `middleware` to be changed, as there can be multiple `middlewares`. So adding this to the `redirect` `middleware` in our case does implement this header.

> (btw, are you on Twitter? I'd love to credit you once I get this in there)

Unfortunately no, but do not worry about that too much. Thanks for your work so far, I'll get back to you once I properly tested the changes.

simonkrenger commented:
Tested with Traefik 2.1, fixed middleware definition in e135a30.

Example output with latest version:
```
# generated 2020-01-17, http://localhost:5500/#server=traefik&server-version=2.1.1&config=intermediate
[http.routers]
  [http.routers.router-secure]
    rule = "Host(`foo.com`)"
    service = "service-id"
    middlewares = ["hsts-header"]

    [http.routers.router-secure.tls]
      options = "intermediate"

  [http.routers.router-insecure]
    rule = "Host(`foo.com`)"
    service = "service-id"
    middlewares = ["redirect-to-https", "hsts-header"]

[http.middlewares]
  [http.middlewares.redirect-to-https.redirectScheme]
    scheme = "https"
  [http.middlewares.hsts-header.headers]
    [http.middlewares.hsts-header.headers.customResponseHeaders]
      Strict-Transport-Security = "max-age=63072000"

# due to Go limitations, it is highly recommended that you use an ECDSA
# certificate, or you may experience compatibility issues
[[tls.certificates]]
  certFile = "/path/to/signed_cert_plus_intermediates"
  keyFile = "/path/to/private_key"

[tls.options]
  [tls.options.intermediate]
    minVersion = "VersionTLS12"
    cipherSuites = [
      "TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384",
      "TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384",
      "TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256",
      "TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256",
      "TLS_ECDHE_ECDSA_WITH_CHACHA20_POLY1305",
      "TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305"
    ]
```

Let me know if there is something else you would like to change. Thanks!

april commented:
A few minor quibbles that I can fix in post-production, but it overall looks great. Thanks for your work on it and for testing it.  :)

april commented:
Should be fully deployed now:
https://ssl-config.mozilla.org/#server=traefik&server-version=2.1.2&config=intermediate

Thanks again!
